### ISSUE TITLE
Unable to connect to the remote server when the server is down and then up

### ISSUE THREAD

**Issue by TheFireCookie** [ExRam/ExRam.Gremlinq#99]
**Describe the bug**

1. Configure Gremlinq to work with Janusgraph on a Janusgraph host that is not available.
2. Do some calls with the configured QuerySource, you get some "Unable to connect to the remote server" exception, this seems normal and expected.
3. Start your janusgraph host and wait for it to be up.
4. Do some calls with the same configured QuerySource (not a new one), The calls still returns the exception "Unable to connect to the remote server" even if the server is up.

**Expected behavior**
The configured querysource should work when the Gremlin server become up and available.

**Version information**
ExRam.Gremlinq: 8.0.0-preview-0707

**Minimal working example**
https://github.com/TheFireCookie/GremlinqBug
Here is the repo to reproduce the repo.
The way to reproduce it:
1. Launch the console with the janusgraph server/docker/the way you want down.
2. Wait for it to show that it can't connect to Janusgraph
3. Start janusgraph
4. Nothing change.
5. Restart the console and it works

**Comment by TheFireCookie** [comment]
Hello @danielcweber , the issue is still the same as before. Maybe I need to forward this bug to the Gremlin.Net repo?

**Comment by danielcweber** [comment]
@FlorianHockmann Is there still a known issue in Gremlin.NET 3.4.8 that would prevent it from connecting if the server is down entirely initially?

**Comment by FlorianHockmann** [comment]
The reconnection mechanism of the Gremlin.NET driver is meant for cases where an existing connection gets closed because the server is temporary unavailable. The case described here is that the server is unavailable when the `GremlinClient` is instantiated. The driver will throw a `WebSocketException` (as described in the issue description) in that case as it cannot reach the specified server. We don't know the cause for not being able to reach the server. It could of course be that the server is simply not available at that very moment but it becomes available shortly after. In that case a simple retry mechanism might sound like a good option. It could however also be that the server is simply never reachable with that config, for example because the given address is simply wrong or because there is a general network problem preventing the driver from reaching the server, like a missing firewall rule.
If it's not a transient problem, then we need to make the user aware of it so they can fix it. That's why I think that it's correct for a driver to throw an exception if the initial connection establishment fails. I would also expect that most often where this fails, it's caused by a general problem and not a transient one.
So, long story short: I don't see a bug here in Gremlin.NET as that is the behaviour that I would expect from a driver. If you guys disagree, then I'm happy to hear your reasoning.

If you expect JanusGraph Server not to be available when you start your client application, then I'd recommend to implement a retry mechanism in your application.

**Comment by danielcweber** [comment]
@FlorianHockmann I think what the OP describes is rather that Gremlin.NET won't connect even when the server becomes reachable again (after a failed first connection attempt). I haven't tried to reproduce it though.

Conceptually, this shouldn't be any different from (connect successfully) -> (lose a connection) -> (reconnect).

@TheFireCookie Have you confirmed that the server is indeed reachable if it's up and running from the very beginning?

**Comment by TheFireCookie** [comment]
So @FlorianHockmann & @danielcweber ,
I redid a test to check the behaviors.
**Case 1:**
1. I start Janusgraph & wait for it to be up
2. I start my console to check the connection. It tells me it's OK.
3. I stop Janusgraph, the console tells me it's not ok (exact message : No connection to the server available which most likely means that the server is completely unavailable.)
4. I restart Janusgraph, the console keeps telling me it's not ok (with the same message, event after 5 minutes).

**Case 2**
1. I start my console to check the connection, it tells me it's not OK since I haven't started Janusgraph yet.
2. I start Janusgraph & the console with the connection retry will never connect neither.

The case I really reproduce in my production code in the case 2. Why?
Because in production, our services runs in docker containers (janusgraph & the service connection to janusgraph using GremlinQ). But since it's docker containers, I don't know which one will be started first and be available.
But anyway I have a work-around. I've made a Gremlin waiter that check the connection by recreating a new gremlin source each time until the connection is good and then my app start correctly.

Matthias.

**Comment by danielcweber** [comment]
All I can add to this at this time is that ExRam.Gremlinq doesn't span up any websocket-connections, nor does it include any-reconnect logic whatsoever. So @TheFireCookie would you be able to reproduce this using just raw Gremlin.NET?

**Comment by TheFireCookie** [comment]
@danielcweber No issue at all with Gremlin.NET alone. It works in the two cases.

**Comment by danielcweber** [comment]
Could you post a sample project that shows that a raw Gremlin.NET client can indeed be reused?

**Comment by TheFireCookie** [comment]
@danielcweber I've updated the solution here with the 2 solutions: https://github.com/TheFireCookie/GremlinqBug

**Comment by danielcweber** [comment]
Haven't reviewed it, but it's still referencing an older preview of Gremlinq...

**Comment by TheFireCookie** [comment]
Yes because I'm using Janusgraph.Net at the same time that does not work with Gremlin.Net 3.4.8 that is bundle with the last preview of GremlinQ.

**Comment by danielcweber** [comment]
Plus, the test doesn't seem to use Gremlin.NET and Gremlinq in the same manner, because you are reusing the GremlinQuerySource but you recreate a fresh Gremlin.NET client every time.

We assume that the Gremlinq.NET client is faulted for good when it can't connect. A GremlinQuerySource doesn't recreate a Gremlin.NET client internally.

**Comment by danielcweber** [comment]
> Yes because I'm using Janusgraph.Net at the same time that does not work with Gremlin.Net 3.4.8 that is bundle with the last preview of GremlinQ.

Well, it is known that Gremlin.NET fixes some reconnection issues in the latest version, what's the point in NOT using it? If you want the latest Gremlin.NET, use the latest Gremlinq.

**Comment by TheFireCookie** [comment]
@danielcweber If you checkout the previous commit on my repo, I was in the last version for GremlinQ and I had the bug described. Here I made a sample that works with the 2 projects to help you and to give you a simple example, but whatever.

**Comment by danielcweber** [comment]
I'm sorry, I can't find the commit that has ExRam.Gremlinq 8.0.0-preview-0926 installed.

Also, try to re-use the raw Gremlin.NET client, because that's what Gremlinq does if you cache the GremlinQuerySource. Otherwise the assumption that raw Gremlin.NET works when Gremlinq doesn't is just not valid.
